src/environment.js

```diff
@@ -56,21 +56,24 @@ function deriveGetBinding(realm: Realm, binding: Binding) {
   return realm.generator.deriveAbstract(types, values, [], createOperationDescriptor("GET_BINDING", { binding }));
 }
 
-export function havocBinding(binding: Binding): void {
-  let realm = binding.environment.realm;
+export function materializeBinding(realm: Realm, binding: Binding): void {
+  let realmGenerator = realm.generator;
   let value = binding.value;
+  if (value !== undefined && realmGenerator !== undefined && value !== realm.intrinsics.undefined)
+    realmGenerator.emitBindingAssignment(binding, value);
+}
+export function leakBinding(binding: Binding): void {
+  let realm = binding.environment.realm;
   if (!binding.hasLeaked) {
     realm.recordModifiedBinding(binding).hasLeaked = true;
-    if (value !== undefined) {
-      let realmGenerator = realm.generator;
-      if (realmGenerator !== undefined && value !== realm.intrinsics.undefined)
-        realmGenerator.emitBindingAssignment(binding, value);
-      if (binding.mutable === true) {
-        // For mutable, i.e. non-const bindings, the actual value is no longer directly available.
-        // Thus, we reset the value to undefined to prevent any use of the last known value.
-        binding.value = undefined;
-      }
-    }
+    materializeBinding(realm, binding);
+  }
+
+  // Havoc the binding
+  if (binding.mutable === true) {
+    // For mutable, i.e. non-const bindings, the actual value is no longer directly available.
+    // Thus, we reset the value to undefined to prevent any use of the last known value.
+    binding.value = undefined;
```

Comment:
It seems to me that this value should be set to an abstract value that has an expression that does a runtime read from the binding. The abstract value should behave like Top otherwise.

I expect that this logic happens elsewhere. It seems to me, however, that this is the right place to put it.